### ISSUE TITLE
R5 SNP/HTTP error "Listener xxx couldn't determine message type!" when text too long

### ISSUE THREAD

**Issue by sleprevost** [fullphat/snarl#26]
Hi, I encounter the "Transport Warning" message described in the issue title.

As far as I see it occurs when the "&text" parameter is too long. What's the max length for this parameter ? Or for the whole request perhaps ?

**Comment by fullphat** [comment]
Which protocol are you using?  Also, any chance you can post the request you're sending?

**Comment by sleprevost** [comment]
Using transport SNP over HTTP, classic API, this query : http://192.168.xxx.yyy:81/version
gives this reply : SNP/2.1/0/OK/47

It really depends on the length

This one works :

http://192.168.xxx.yyy:81/notify?title=%5BTEST%40vmpfc02%5D%20-%20TEST%20%28PID%205722%29&text=24-04-2017%2009%3A26%3A59%0D%0AIncenderat%20autem%20audaces%20usque%20ad%20insaniam%20homines%20ad%20haec%2C%20quae%20nefariis%20egere%20conatibus%2C%20Luscus%20quidam%20curator%20urbis%20subito%20visus%3A%20eosque%20ut%20heiulans%20baiolorum%20praecentor%20ad%20expediendum%20quod%20orsi%20sunt%20incitans%20vocibus%20crebris.%20qui%20haut%20longe%20postea%20ideo%20vivus%20exustus%20est&icon=%21system-blocked&uid=TEST%3A%3ATEST%3A%3Avmpfc02&priority=2&action=View%20history,http%3A%2F%2Flocalhost%3A9989%2Fhistory

This one fails (same parameters but longer text) :

http://192.168.231.173:81/notify?title=%5BTEST%40vmpfc02%5D%20-%20TEST%20%28PID%205722%29&text=24-04-2017%2009%3A32%3A45%0D%0AIncenderat%20autem%20audaces%20usque%20ad%20insaniam%20homines%20ad%20haec%2C%20quae%20nefariis%20egere%20conatibus%2C%20Luscus%20quidam%20curator%20urbis%20subito%20visus%3A%20eosque%20ut%20heiulans%20baiolorum%20praecentor%20ad%20expediendum%20quod%20orsi%20sunt%20incitans%20vocibus%20crebris.%20qui%20haut%20longe%20postea%20ideo%20vivus%20exustus%20est%20plus%20long%20plus%20long%20plus%20long%20plus%20plus%20plus%20plus%20plus%20plus%20encore%20plus%20beaucoup%20plus%20tres%20tres%20long%20et%20bien%20plus%20encore%20c%20est%20étrange%20j%20ai%20du%20mal%20a%20faire%20suffisamment%20long%20c%20est%20penible%20ca%20va%20durer%20encore%20longtemps&icon=%21system-blocked&uid=TEST%3A%3ATEST%3A%3Avmpfc02&priority=2&action=View%20history,http%3A%2F%2Flocalhost%3A9989%2Fhistory

I'm aware a notification can't of course be a whole book ;-), I'm just checking the max length to take care of it in my script and handle this case instead of calling snarl with a too long text argument.

**Comment by fullphat** [comment]
Thanks - will test it out against R5 Beta 2.  Which version of Snarl are you using?  There's no specific limit on request length, especially when it comes to encoded data.  R5 Beta 2 brings massive improvements to all networking protocols so I'll make sure this one works.

**Comment by sleprevost** [comment]
I'm using 5.0 Beta 1 (V47.89)

**Comment by sleprevost** [comment]
With a few more tests, I saw that :
- Using curl, if the complete URL is in the range 0..831 chars it works. 832 or more chars fails
- Using IE, the range is from 0 to 591. 592 chars and more fails
- Using IE with debug activated (F12 key) : range is 0 to 567.

So... I suspected that the browser header size differs... and I was right !

Using the curl trace I see that :

831 chars in url  + curl header = 1024 chars => OK
832 chars in url + curl header = 1025 chars => FAILED

Looks like there a limit of 1024 chars somewhere, don't know if it's on HTTP or on SNP server side. But I hope it will help you ;-)

**Comment by fullphat** [comment]
Interesting!  I've not had chance to test yet, but maybe it's an issue with GET requests in general.  As a side note, R5 Beta 2 will include a new version of snp/http (v2) which uses a RESTful approach.  This might avoid the issue entirely as you create notifications by POSTing a JSON request - see docs here: https://github.com/fullphat/snarl/wiki/SNP-HTTP-API

**Comment by fullphat** [comment]
Tested and working in Beta 2!  Let's just say it's a very good thing I'm now using the .net HttpListener object and not my own hacked-together HTTP-using-a-socket code that was in Beta 1. :man_facepalming:

**Comment by sleprevost** [comment]
:+1: 

Nice ! So 1024 bytes won't be anymore a limit in beta 2 ?
According to the roadmap, when do you plan to deliver it ?

**Comment by fullphat** [comment]
Correct.  So far as getting Beta 2 out goes, let me see if I can get it (or a Dev Release) out sooner than planned.  Been doing some work on settings management so need to check I've not totally broken that. 😄

**Comment by fullphat** [comment]
I've uploaded R5.0 Beta 2 Preview A here: https://sourceforge.net/projects/snarlwin/files/Snarl/R5/ if you want to give it a try?

**Comment by sleprevost** [comment]
I've just installed and tested it.

2 points :

- I can't use 81 port anymore (doesn't listen) but works on anaother port (8080 for example)
- I can't anymore curl the URL using IP adress (looks like a name is mandatory)

![2017-04-28 09_38_05-bad request](https://cloud.githubusercontent.com/assets/26896051/25519116/d3cdcccc-2bf6-11e7-87c7-45b2b65c0415.jpg)

Anyway very long messages works (over 2048 bytes) but only from a web browser and using "localhost" (even IP 127.0.0.1 doesn't work :-( )

**Comment by fullphat** [comment]
.net does seem funny about low-numbered ports.  Try adding an ACL on the machine if you want to use port 81.  You'll also need to add ACLs to allow use of IP addresses too.  Try this:

`netsh http add urlacl url=http://<ip>:81/ user=Everyone listen=yes`

Does that work?

**Comment by sleprevost** [comment]
Port 81 is not a issue for me : we plan to use 8080 when deploying in production mode and it works from my browser using "localhost". I was just notifying you a slight difference with the previous beta ;-)

However, not being able to call snarl by IP can be an issue because I'm sending notifications from Linux servers (using curl) in a subnet  to windows desktops in another subnet, and servers don't (and don't have to) resolve names for our desktops :-(

**Comment by sleprevost** [comment]
Failed... :-(
Reply (translated from French) is : SDDL creation failed : 1332

**Comment by sleprevost** [comment]
Ok, "Everyone" had to be translated in French ;-)
So, I added the ACL and snarl cannot listen on port 8080... I deleted it... and it listen again.
Still working with "localhost" aftre deleted it but still not with the IP address

**Comment by fullphat** [comment]
Ok.  I've got the low port issue resolved (although you don't need that now):

`netsh http add iplisten ipaddress=0.0.0.0:81`

Will look at the IP address/hostname issue next...

**Comment by fullphat** [comment]
> Still working with "localhost" aftre deleted it but still not with the IP address

I've found the error.  It's a bug in the SNP/HTTP listener which I've now fixed.  I'll try to a get Beta 2 Rel. B out later today but it might be after the weekend now.  I can confirm the issue is resolved though. 😄

**Comment by sleprevost** [comment]
Good news ! And so fast !
I'll be off from tonight to may the 9th, so you've got time 

Anyway, I'll be happy to continue to help you improve release 5 when I'll be back.

For information I plan to use snarl in two ways :
- Notifications from different servers when something goes wrong in some batch process
- Notifications from nagios server (on critcal hosts and services only)

Each of them uses a bash script calling curl to request snarl. Next release will have to deal with JSON requests.

**Comment by fullphat** [comment]
No problem! 👍 

Thanks for the continued support and the background - if you're interested in doing a blog post on your work at some point, I'll put it on our site. 😄

**Comment by fullphat** [comment]
Ok - Beta 2 Rel. B uploaded: https://sourceforge.net/projects/snarlwin/files/Snarl/R5/ give it a go! 😀

**Comment by sleprevost** [comment]
Hi, back to work today.

Unfortunately, same thing with beta 2 rel B :

- http://localhost:8080/ work

- http://127.0.0.1:8080/ don't work

**Comment by sleprevost** [comment]
Here is the debug log content :

```
-=-=- Snarl logging started -=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-
bootstrap: it's now: Tue, 09 May 2017 11:50:07 GMT
bootstrap: o/s info follows...
  platform: Win32NT
  version: Microsoft Windows NT 6.1.7601 Service Pack 1
  service pack: Service Pack 1
bootstrap: server info follows...
  release: 5.0 Beta 2 (Rel B)
  api: 47
  sys: 47.96
bootstrap: initialising settings...
bootstrap: starting UI...
bootstrap: checking user data folders...
bootstrap: %APPDATA% is "C:\Users\s.leprevost\AppData\Roaming\"
bootstrap: starting Win32 IPC transport...
bootstrap: starting management server on '*:9989'...
bootstrap: didn't start: falling back to localhost...
bootstrap: management server started!
bootstrap: starting SnarlOSD...
bootstrap: starting application roster...
bootstrap: starting extension roster...
bootstrap: starting rules roster...
bootstrap: starting presence monitor...
bootstrap: enabling...
bootstrap: displaying welcome message...
bootstrap: starting network roster...
bootstrap: main startup completed, other threads may still be initialising
$$name/msgx
$$port/8080
$$protocol/snphttp
$$password/
$$allow_forwards/always
```

**Comment by fullphat** [comment]
If you run

`netsh http show urlacl`

Is there a specific entry for http://127.0.0.1:8080?

I've got it working on port 81 (from previous testing) and when I run the command I get (amongst others) this:

```
    Reserved URL            : http://127.0.0.1:81/
        User: \Everyone
            Listen: Yes
            Delegate: No
            SDDL: D:(A;;GX;;;WD)
```

**Comment by sleprevost** [comment]
After a few more tests :

- If I run snarl.exe as Administrator, all works fine
- If I run as "myself", same problem (listening OK but expext an explicit hostname)
- If I run as "myself" WITH urlacl : snarl doesn't listen

Running snarl as an administrator isn't issue for me but as far as I see if I use snarl beta 2 :

- I must run as an administrator (wasn't mandatory in beta 1)
- Calls by IP adress aren't allowed if not admistrator (by name calls works)
- I can't see icons anymore in notifications (does I need to provide a path ?)

**Comment by fullphat** [comment]
Ok - try Beta 2 Rel C (just uploaded 😄)

I've managed to replicate what you did and it's working for me using Rel C.  Here's what I did:

* Used a clean Windows 7 VM
* Created a non-admin user
* Installed Rel C
* Launched Snarl
* Created an SNP/HTTP listener on port 8081
* Launched an Administrator Command Prompt (and entered admin user password)
* Created a URLACL reservation: `netsh http add urlacl http://+:8081/ user=Everyone listen=ye
s`
* Quit Snarl
* Launched Snarl
* Launched IE
* Browsed to `http://localhost:8081/` - worked
* Browsed to `http://127.0.0.1:8081/` - worked
* Browsed to `http://192.168.1.77:8081/` - worked

Can you try the same?

**Comment by sleprevost** [comment]
Ok, here are the results : 

- Deleted %APPDATA%\Roaming\full phat
- Installed R5 Beta 2 Rel C
- Launched Snarl (as non admin user)
- Created SNP/HTTP listener on port 8081 : ok, listening
- Browsed to http://localhost:8080/ : ok
- Browsed to http://127.0.0.1:8080/ : ok
- Browsed to http://192.168.xxx.yyy:8080/ : ok

All works fine, no urlacl had to be set up !! Thanks a lot !

On icon side, my last "trick" (changing for full path) which worked with Rel B doesn't work anymore : the grampf icon appears, even if I set a "icon" parameter.

To sum up on icon side : 

- Beta 1 : `icon=!system-blocked.png` works
- Beta 2 Rel B : `icon=etc%2Ficons%2Fstock%2Fsystem-blocked.png` works
- Beta 2 Rel C : none of those two syntaxes works

**Comment by fullphat** [comment]
Excellent!

On the icon side of things, I have been doing some messing around so it looks like I've broken something.  Let me check...

**Comment by fullphat** [comment]
Further update... 😄 

SNP/HTTP wasn't url-decoding the request so that was causing file paths to get messed up.  The following now both work in Rel. D (note the differences around passing the icon extension):

`curl "http://127.0.0.1:8080/notify?title=Hello,+World!&icon=etc%2Ficons%2Fstock%2Fsystem-blocked.png"`

`curl "http://127.0.0.1:8080/notify?title=Hello&icon=!misc-chair"`

As a general point, using stock icons by prefixing them with `!` is preferred as Snarl 5.0 will also pass the URL to the web-hosted variant of the stock icon.  This means web-based services such as Boxcar can still display the same icon.

**Comment by sleprevost** [comment]
Ok, just tried with Rel C using `!system-blocked` instead of `%21system-blocked` and it works.
I also prefer using '!' : shorter URL, no need to specify any path. Using full path was only a trick I used since Beta 2 Rel B.

To sum up improvements and bugs fixed since beta 1 :
- No more 1024 bytes limit on the request size using SND/HTTP
- Request works with IP address and not only by name
- Full icons path no more mandatory

Good job ! :+1: :1st_place_medal:

**Comment by fullphat** [comment]
Awesome!  I promise to try not to break it again now 😄 

As for UIDs, that's still a work-in-progress, properties such as `replace-uid`, `merge-uid`, `update-uid`, are currently either broken or may generate odd effects.  I'll take a look at the specific issue with `uid` first though. 👍

**Comment by fullphat** [comment]
Closing...
